File: internal/i18n/catalog_test.go
package i18n

import (
	"reflect"
	"sort"
	"strings"
	"testing"
)

// The second mechanism (hard rule 9).
//
// The `Catalog` type is the first, and in Go it is weaker than TypeScript's
// `Copy`: both catalogs are the same type whatever either of them left out,
// because Go has a zero value. `Missing` closes that, and everything below
// closes the holes `Missing` cannot see — an English leaf that is a verbatim
// copy of the Indonesian, a format verb that survived into the output, a
// sentence that scolds.

func bothCatalogs() map[Locale]*Catalog {
	return map[Locale]*Catalog{ID: &idCatalog, EN: &enCatalog}
}

func TestEveryLocaleHasACatalog(t *testing.T) {
	// A Locale added to `Locales` with no catalog behind it would be served by
	// `For` as Indonesian, silently. FromContext's contract is that what comes
	// back is always a locale with a catalog behind it; this is the half of
	// that promise this file owns.
	for _, l := range Locales {
		if _, ok := catalogs[l]; !ok {
			t.Errorf("Locales carries %q with no catalog behind it", l)
		}
	}
	if len(catalogs) != len(Locales) {
		t.Errorf("catalogs holds %d entries, Locales declares %d", len(catalogs), len(Locales))
	}
}

func TestNoLeafIsUnwritten(t *testing.T) {
	// This is the missing-key check, and it runs in both directions by
	// construction: every catalog is the same struct, so a key present in one
	// and absent in the other shows up here as the absent one being empty.
	for l, c := range bothCatalogs() {
		if missing := Missing(c); len(missing) > 0 {
			t.Errorf("%s.go has %d unwritten leaves:\n  %s",
				l, len(missing), strings.Join(missing, "\n  "))
		}
	}
}

func TestForNeverReturnsNil(t *testing.T) {
	for _, l := range append([]Locale{"", "pt-BR", "EN"}, Locales...) {
		if For(l) == nil {
			t.Errorf("For(%q) returned nil", l)
		}
	}
	if For("pt-BR") != For(Default) {
		t.Error("an unknown locale must fall back to the default catalog")
	}
}

func TestNothingIsLeftUntranslated(t *testing.T) {
	// `Missing` cannot see the likeliest way `en.go` goes wrong: a leaf
	// copy-pasted from `id.go` and never touched. It has a written value, it
	// has the right type, and it renders Indonesian to an English reader.
	//
	// Nothing in this catalog is legitimately identical across the two
	// languages — every leaf is a whole sentence. If that ever stops being
	// true, add the leaf here with the reason rather than deleting the test.
	sameOnPurpose := map[string]string{}

	same := identical(reflect.ValueOf(idCatalog), reflect.ValueOf(enCatalog))
	var bad []string
	for _, path := range same {
		if _, ok := sameOnPurpose[path]; !ok {
			bad = append(bad, path)
		}
	}
	sort.Strings(bad)

	if len(bad) > 0 {
		t.Errorf("identical in id.go and en.go, so one of them was never translated:\n  %s",
			strings.Join(bad, "\n  "))
	}
}

func TestNoFormatVerbSurvivesIntoOutput(t *testing.T) {
	// A `%d` on the screen means a Sprintf lost an argument in translation,
	// and `%!d(MISSING)` means it lost the argument entirely. Both compile.
	for l, c := range bothCatalogs() {
		for path, s := range renderAll(reflect.ValueOf(*c)) {
			if strings.Contains(s, "%!") {
				t.Errorf("%s.go %s: Sprintf reported a bad argument: %q", l, path, s)
			}
			// The probe calls every func with zero values, so a surviving verb
			// is a verb the format string never consumed.
			for _, verb := range []string{"%d", "%s", "%v", "%q"} {
				if strings.Contains(s, verb) {
					t.Errorf("%s.go %s: unconsumed %s in %q", l, path, verb, s)
				}
			}
		}
	}
}

func TestNeverPunitive(t *testing.T) {
	// Hard rule 6, and the same list `web/src/i18n/catalog.test.ts` uses,
	// because the failure is the same one. Not a complete list of ways to
	// blame somebody — there is no such list — which is why the paragraph at
	// the top of `en.go` is the real instruction and this is the backstop.
	banned := map[Locale][]string{
		EN: {
			"don't forget", "do not forget", "remember to", "make sure you",
			"you missed", "you forgot", "you should have", "you failed",
			"you fell behind", "falling behind", "behind on", "overdue",
			"streak", "keep it up", "stay on track", "don't break",
			"oops", "uh oh", "!",
		},
		ID: {"jangan lupa", "kamu lupa", "kamu gagal", "jangan sampai", "seharusnya", "beruntun", "!"},
	}

	for l, c := range bothCatalogs() {
		for path, s := range renderAll(reflect.ValueOf(*c)) {
			lower := strings.ToLower(s)
			for _, phrase := range banned[l] {
				if strings.Contains(lower, phrase) {
					t.Errorf("%s.go %s: %q — %q", l, path, phrase, s)
				}
			}
		}
	}
}

func TestGroupFormatsForTheLocaleThatAsked(t *testing.T) {
	// 07 L8's quotas are 5.000 notes and 20.000 cards. A bare strconv.Itoa
	// prints 5000, which is wrong in both languages.
	cases := []struct {
		n    int
		sep  string
		want string
	}{
		{0, dot, "0"},
		{999, dot, "999"},
		{1000, dot, "1.000"},
		{5000, dot, "5.000"},
		{20000, comma, "20,000"},
		{300, comma, "300"},
		{1234567, comma, "1,234,567"},
		{-5000, dot, "-5.000"},
	}
	for _, tc := range cases {
		if got := group(tc.n, tc.sep); got != tc.want {
			t.Errorf("group(%d, %q) = %q, want %q", tc.n, tc.sep, got, tc.want)
		}
	}

	// And the catalogs actually use it, rather than each having reached for
	// strconv on its own.
	if got := idCatalog.Quota.Notes(5000); !strings.Contains(got, "5.000") {
		t.Errorf("Indonesian quota message does not group with a dot: %q", got)
	}
	if got := enCatalog.Quota.Notes(5000); !strings.Contains(got, "5,000") {
		t.Errorf("English quota message does not group with a comma: %q", got)
	}
}

func TestMissingReportsTheLeafThatIsBlank(t *testing.T) {
	// The mechanism needs its own test, or it is a hope that a test is
	// running. A catalog with two holes in it must name both.
	var hollow Catalog
	hollow.Common.BadRequest = "ada"

	missing := Missing(&hollow)
	if len(missing) == 0 {
		t.Fatal("Missing reported nothing for an empty catalog")
	}
	if contains(missing, "Common.BadRequest") {
		t.Error("Missing reported a leaf that was written")
	}
	for _, want := range []string{"Common.NotFound", "Common.ServerErrorWithCode", "Quota.Notes"} {
		if !contains(missing, want) {
			t.Errorf("Missing did not report %s", want)
		}
	}

	// A func that returns nothing is as blank as a field nobody filled in.
	blank := Catalog{}
	blank.Auth.PasswordTooShort = func(int) string { return "  " }
	if !contains(Missing(&blank), "Auth.PasswordTooShort") {
		t.Error("Missing did not report a func that produces an empty string")
	}
}

// renderAll walks a catalog and returns every string it can produce, keyed by
// path. Funcs are called with zero values, which is enough to see the sentence
// around the argument.
func renderAll(v reflect.Value) map[string]string {
	out := map[string]string{}
	render(v, "", out)
	return out
}

func render(v reflect.Value, path string, out map[string]string) {
	switch v.Kind() {
	case reflect.Struct:
		t := v.Type()
		for i := 0; i < v.NumField(); i++ {
			render(v.Field(i), join(path, t.Field(i).Name), out)
		}
	case reflect.String:
		out[path] = v.String()
	case reflect.Func:
		if s, ok := probe(v); ok {
			out[path] = s
		}
	}
}

// identical returns the paths whose rendered value is the same in both
// catalogs.
func identical(a, b reflect.Value) []string {
	left, right := renderAll(a), renderAll(b)
	var same []string
	for path, s := range left {
		if other, ok := right[path]; ok && s == other {
			same = append(same, path)
		}
	}
	return same
}

func contains(haystack []string, needle string) bool {
	for _, s := range haystack {
		if s == needle {
			return true
		}
	}
	return false
}

File: web/src/i18n/catalog.test.ts
import { describe, expect, it } from 'vitest'
import { en } from './en'
import { id } from './id'
import { bootLocale, rememberLocale } from './boot'
import { catalogLoaded, copyFor, DEFAULT_LOCALE, isLocale, loadCatalog, LOCALES } from './index'
import { pluralFor } from './plural'

/**
 * The second mechanism (hard rule 9).
 *
 * The `Copy` type is the first: a key in one catalog and not the other does not
 * compile. That is strong, and it has two holes this file covers.
 *
 * The first is optionality. The moment a field is declared `foo?: string` —
 * and the landing repo's `Feature.caption` shows how reasonable that looks —
 * the type stops requiring both catalogs to have it, and a screen renders a
 * caption in Indonesian and nothing in English. Parity is checked here by
 * walking the objects, so it does not care what the type made optional.
 *
 * The second is tone. `en.ts` opens with a paragraph about the vocabulary of
 * gentle blame that English has and Indonesian does not, and a paragraph is
 * a hope. This is the mechanism.
 */

type Node = unknown

/**
 * Both catalogs, for the tests that need to walk them.
 *
 * Built here rather than imported from `index.tsx`, which no longer exports a
 * map: `en` is behind a dynamic import so that it stays out of the entry chunk,
 * and a runtime map holding it would defeat that. A test file is not bundled,
 * so it may import both directly.
 */
const BOTH: Record<string, Node> = { id, en }

/** Every leaf, as `path -> kind`. Functions are probed for the strings inside. */
function walk(node: Node, path = '', out = new Map<string, string>()) {
  if (typeof node === 'string') {
    out.set(path, 'string')
    return out
  }

  if (typeof node === 'function') {
    // Arity is part of the contract: a translation that drops an argument
    // still typechecks if it ignores it, and then prints a sentence with the
    // value missing.
    out.set(path, `function/${(node as (...a: never[]) => string).length}`)
    return out
  }

  if (Array.isArray(node)) {
    out.set(path, `array/${node.length}`)
    node.forEach((child, i) => walk(child, `${path}[${i}]`, out))
    return out
  }

  if (node && typeof node === 'object') {
    for (const [key, child] of Object.entries(node)) {
      walk(child, path ? `${path}.${key}` : key, out)
    }
    return out
  }

  out.set(path, typeof node)
  return out
}

/**
 * Every string a catalog can produce, including the ones built inside a
 * function. Functions are called with probe arguments rather than read as
 * source, so a plural's `one` and `other` forms are both seen.
 */
function allStrings(node: Node, out: string[] = []): string[] {
  if (typeof node === 'string') {
    out.push(node)
    return out
  }

  if (typeof node === 'function') {
    const fn = node as (...args: unknown[]) => unknown
    // 1 and 2 cover `one` and `other` in every locale this app ships. The
    // string probe is for the functions that interpolate a name rather than a
    // count; passing a number to those is harmless.
    for (const probe of [1, 2, 0, 'x']) {
      try {
        const result = fn(...Array.from({ length: fn.length }, () => probe))
        if (typeof result === 'string') out.push(result)
      } catch {
        // A function that cannot be probed is not evidence of anything.
      }
    }
    return out
  }

  if (node && typeof node === 'object') {
    for (const child of Object.values(node)) allStrings(child, out)
  }

  return out
}

describe('catalog parity', () => {
  const idLeaves = walk(id)
  const enLeaves = walk(en)

  it('has the same keys in both directions', () => {
    const missingFromEn = [...idLeaves.keys()].filter((k) => !enLeaves.has(k))
    const missingFromId = [...enLeaves.keys()].filter((k) => !idLeaves.has(k))

    expect(missingFromEn, 'in id.ts and not in en.ts').toEqual([])
    expect(missingFromId, 'in en.ts and not in id.ts').toEqual([])
  })

  it('agrees on the shape of every key', () => {
    // A string in one catalog and a function in the other means one of them
    // stopped interpolating, which renders as a literal `#` on screen. A
    // tuple of three in one and two in the other renders as a short list.
    const mismatched = [...idLeaves.entries()]
      .filter(([key, kind]) => enLeaves.has(key) && enLeaves.get(key) !== kind)
      .map(([key, kind]) => `${key}: id=${kind} en=${enLeaves.get(key)}`)

    expect(mismatched).toEqual([])
  })

  it('has no blank strings', () => {
    // An empty string typechecks and reads as a missing translation nobody
    // reported, because there is nothing on screen to report.
    for (const [locale, catalog] of Object.entries(BOTH)) {
      const blank = allStrings(catalog).filter((s) => s.trim() === '')
      expect(blank, `blank strings in ${locale}.ts`).toEqual([])
    }
  })

  it('leaves no unreplaced placeholder', () => {
    // `#` survives into the output when a form is written without it.
    for (const [locale, catalog] of Object.entries(BOTH)) {
      const stray = allStrings(catalog).filter((s) => s.includes('#'))
      expect(stray, `unreplaced # in ${locale}.ts`).toEqual([])
    }
  })
})

describe('never punitive (hard rule 6)', () => {
  /**
   * The phrases that keep coming back. Not a complete list of ways to blame
   * someone in English — there is no such list — which is why the paragraph at
   * the top of `en.ts` is the real instruction and this is the backstop.
   */
  const ENGLISH = [
    "don't forget",
    'do not forget',
    'remember to',
    'make sure you',
    'you missed',
    'you forgot',
    'you fell behind',
    'falling behind',
    'behind on',
    'streak',
    'keep it up',
    'stay on track',
    "don't break",
    'oops',
    'uh oh',
    '!',
  ]

  const INDONESIAN = ['jangan lupa', 'kamu lupa', 'jangan sampai', 'beruntun', '!']

  const cases: { locale: string; catalog: Node; banned: string[] }[] = [
    { locale: 'en', catalog: en, banned: ENGLISH },
    { locale: 'id', catalog: id, banned: INDONESIAN },
  ]

  for (const { locale, catalog, banned } of cases) {
    it(`${locale}.ts uses none of the punitive phrases`, () => {
      const found = allStrings(catalog).flatMap((line) =>
        banned
          .filter((phrase) => line.toLowerCase().includes(phrase))
          .map((phrase) => `${phrase} — ${line}`),
      )

      expect(found).toEqual([])
    })
  }
})

describe('plural', () => {
  it('selects one and other in English', () => {
    const n = pluralFor('en')
    const forms = { one: '# card', other: '# cards' }

    expect(n(0, forms)).toBe('0 cards')
    expect(n(1, forms)).toBe('1 card')
    expect(n(2, forms)).toBe('2 cards')
  })

  it('falls back to other for a category a locale did not write', () => {
    // Indonesian has one form. Writing six would be six copies of the same
    // sentence, so `other` answers everything.
    const n = pluralFor('id')
    expect(n(1, { other: '# kartu' })).toBe('1 kartu')
    expect(n(2, { other: '# kartu' })).toBe('2 kartu')
  })

  it('formats the number for the locale', () => {
    // 07 L8's quotas are 5.000 notes and 20.000 cards. A raw `${n}` prints
    // 5000 in both languages and is wrong in both.
    expect(pluralFor('id')(5000, { other: '# catatan' })).toBe('5.000 catatan')
    expect(pluralFor('en')(5000, { one: '# note', other: '# notes' })).toBe('5,000 notes')
  })

  it('replaces every occurrence of the placeholder', () => {
    expect(pluralFor('en')(2, { other: '# of #' })).toBe('2 of 2')
  })
})

describe('the accessor', () => {
  it('lists Indonesian first, and lists the fallback', () => {
    expect(LOCALES).toEqual(['id', 'en'])
    expect(LOCALES).toContain(DEFAULT_LOCALE)
  })

  it('recognises only the locales that exist', () => {
    // I2 resolves a locale off a user setting and an Accept-Language header,
    // both of which are user input. `pt-BR` must not index the catalog.
    expect(isLocale('id')).toBe(true)
    expect(isLocale('en')).toBe(true)
    expect(isLocale('pt-BR')).toBe(false)
    expect(isLocale(undefined)).toBe(false)
  })

  it('falls back to Indonesian, which is the source language', () => {
    expect(DEFAULT_LOCALE).toBe('id')
  })
})

describe('the lazy split', () => {
  // English is behind `import()` so I5's two languages do not both land in the
  // chunk a signed-out stranger waits on. These assert the properties that buys
  // and the one it costs.

  it('has Indonesian in memory without loading anything', () => {
    // The fallback can never be a network request (hard rule 8): if it were,
    // a dropped connection would render nothing at all rather than Indonesian.
    expect(catalogLoaded('id')).toBe(true)
    expect(copyFor('id')).toBe(id)
  })

  it('falls back to Indonesian for a locale that has not arrived', () => {
    // Not a papered-over failure — it is the documented fallback, and it is
    // why a failed chunk fetch is a language downgrade rather than a blank page.
    expect(copyFor('en')).toBe(id)
  })

  it('serves English once its chunk is loaded', async () => {
    await loadCatalog('en')
    expect(catalogLoaded('en')).toBe(true)
    expect(copyFor('en')).toBe(en)
  })

  it('loads a chunk once, however many callers ask', async () => {
    // StrictMode mounts, unmounts and remounts every effect in dev, so the
    // provider asks twice for one switch. Sharing the in-flight promise is what
    // keeps that from being two fetches.
    const [a, b] = await Promise.all([loadCatalog('en'), loadCatalog('en')])
    expect(a).toBe(b)
    expect(a).toBe(en)
  })
})

describe('the boot hint', () => {
  // The first paint's locale has to be answerable synchronously, because an
  // effect runs after the paint and the wrong language is legible. Same shape
  // and same reason as web/public/theme.js (D-086).
  //
  // The order inside `bootLocale` is cache → browser → id, which is the
  // client's half of D-094's account → Accept-Language → id. The browser step
  // has to be stubbed in every case that is not about it, because jsdom
  // reports an English navigator and would otherwise make "the fallback" and
  // "the browser's answer" indistinguishable.

  /** Run `fn` with `navigator.languages` reporting `tags`. */
  function withBrowserLanguages<T>(tags: string[], fn: () => T): T {
    const original = Object.getOwnPropertyDescriptor(navigator, 'languages')
    Object.defineProperty(navigator, 'languages', {
      configurable: true,
      get: () => tags,
    })
    try {
      return fn()
    } finally {
      if (original) Object.defineProperty(navigator, 'languages', original)
      else delete (navigator as unknown as Record<string, unknown>).languages
    }
  }

  it('answers Indonesian when nothing is remembered and the browser asks for nothing we have', () => {
    expect(withBrowserLanguages(['pt-BR', 'ja'], bootLocale)).toBe('id')
  })

  it('reads back what was remembered', () => {
    rememberLocale('en')
    expect(withBrowserLanguages(['id'], bootLocale)).toBe('en')
  })

  it('ignores a stored value that is not a locale', () => {
    // The cache outlives deploys. A locale removed in a later version must not
    // be able to index the catalog from a browser that still has it.
    localStorage.setItem('konku.locale', 'pt-BR')
    expect(withBrowserLanguages(['pt-BR'], bootLocale)).toBe('id')
  })

  it('survives storage being unavailable', () => {
    // A private window, disabled site data, or a thumbnailer with a hostile
    // getter. None of them is a reason to fail to paint.
    const getItem = Storage.prototype.getItem
    Storage.prototype.getItem = () => {
      throw new Error('denied')
    }
    try {
      expect(withBrowserLanguages(['pt-BR'], bootLocale)).toBe('id')
    } finally {
      Storage.prototype.getItem = getItem
    }
  })

  // The browser step (ticket 11 I2). A stranger's first visit has no cache and
  // no account, so without this their first paint is Indonesian whatever they
  // asked for — and the correction would arrive after the paint, which is the
  // one thing this file exists to prevent.

  it('follows the browser when nothing is remembered', () => {
    expect(withBrowserLanguages(['en-GB', 'en'], bootLocale)).toBe('en')
    expect(withBrowserLanguages(['id-ID'], bootLocale)).toBe('id')
  })

  it('takes the first language it has copy for, not the first listed', () => {
    expect(withBrowserLanguages(['pt-BR', 'ja-JP', 'en-US'], bootLocale)).toBe('en')
  })

  it('lets the cache outrank the browser, because the account wrote it', () => {
    // The cache is written by resolution, so by the second visit it carries
    // the account's own setting — which outranks the browser (D-094).
    rememberLocale('id')
    expect(withBrowserLanguages(['en-US'], bootLocale)).toBe('id')
  })
})
